apps/bmd/src/AppRoot.tsx

```diff
@@ -141,898 +147,1091 @@ export const activationStorageKey = 'activation'
+          const writeIn = tallyContestWriteIns.find(
+            (c) => c.name === candidate.name
+          )
+          if (typeof writeIn === 'undefined') {
+            tallyContestWriteIns.push({
+              name: candidate.name,
+              tally: 1,
+            })
+          } else {
+            writeIn.tally++
+          }
+        } else {
+          const candidateIndex = contest.candidates.findIndex(
+            (c) => c.id === candidate.id
+          )
+          if (
+            candidateIndex < 0 ||
+            candidateIndex >= candidateContestTally.candidates.length
+          ) {
+            throw new Error(
+              `unable to find a candidate with id: ${candidate.id}`
+            )
+          }
+          candidateContestTally.candidates[candidateIndex]++
+        }
+      })
     }
-
-    this.lastVoteUpdateAt = Date.now()
-  }
-
-  public resetVoterData = (): void => {
-    this.props.storage.remove(activationStorageKey)
-    this.props.storage.remove(votesStorageKey)
-  }
-
-  public unconfigure = (): void => {
-    this.setState(this.initialUserState)
-    this.props.storage.clear()
-    this.props.history.push('/')
-  }
-
-  public storeAppState = (): void => {
-    const {
-      appPrecinctId,
-      ballotsPrintedCount,
-      isLiveMode,
-      isPollsOpen,
-      tally,
-    } = this.state
-    this.props.storage.set(stateStorageKey, {
-      appPrecinctId,
-      ballotsPrintedCount,
-      isLiveMode,
-      isPollsOpen,
-      tally,
-    })
   }
+  return tally
+}
 
-  public retrieveAppState = (): Partial<State> => {
-    return this.props.storage.get(stateStorageKey) || {}
+// Sets State. All side effects done outside: storage, fetching, etc
+type AppActions =
+  | { type: 'processAdminCard' }
+  | { type: 'processPollWorkerCard' }
+  | { type: 'processVoterCard'; voterState: Partial<InitialUserState> }
+  | { type: 'pauseCardProcessing' }
+  | { type: 'resumeCardProcessing' }
+  | { type: 'setMachineConfig'; machineConfig: MachineConfig }
+  | { type: 'updateLastVoteUpdateAt'; date: number }
+  | { type: 'unconfigure' }
+  | { type: 'updateVote'; contestId: string; vote: OptionalVote }
+  | { type: 'forceSaveVoteFlag'; value: boolean }
+  | { type: 'resetBallot' }
+  | { type: 'setUserSettings'; userSettings: PartialUserSettings }
+  | { type: 'updateAppPrecinctId'; appPrecinctId: string }
+  | { type: 'enableLiveMode' }
+  | { type: 'toggleLiveMode' }
+  | { type: 'togglePollsOpen' }
+  | { type: 'updateTally' }
+  | { type: 'updateElectionDefinition'; electionDefinition: ElectionDefinition }
+  | { type: 'startWritingLongValue' }
+  | { type: 'finishWritingLongValue' }
+  | { type: 'updateHardwareState'; hardwareState: Partial<HardwareState> }
+  | { type: 'initializeAppState'; appState: Partial<State> }
+  | { type: 'updateLastCardDataString'; currentCardDataString: string }
+
+const appReducer = (state: State, action: AppActions): State => {
```

Comment:
This seems like something that could (eventually) be moved to another file

Reply:
For sure! Eventually!